Improve dynamics section of docs (#2271)
qutip/qutip: doc/guide/dynamics/dynamics-propagator.rst

.. _propagator:

*********************
Computing propagators
*********************

Sometime the evolution of a single state is not sufficient and the full propagator
is desired. QuTiP has the :func:`.propagator` function to compute them:

.. code-block::

    >>> H = sigmaz() + np.pi *sigmax()
    >>> psi_t = sesolve(H, basis(2, 1), [0, 0.5, 1]).states
    >>> prop = propagator(H, [0, 0.5, 1])

    >>> print((psi_t[1] - prop[1] @ basis(2, 1)).norm())
    2.455965272327082e-06

    >>> print((psi_t[2] - prop[2] @ basis(2, 1)).norm())
    2.0071900004562142e-06

> The first argument is the Hamiltonian, any time dependent system format is
> accepted. The function also take an optional `c_ops` for collapse operator,
> when used, propagator for density matrices are computed

Reviewer: ```suggestion
The first argument is the Hamiltonian, any time dependent system format is
accepted. The function also accepts an optional `c_ops` argument for collapse operators.
When used, a propagator for density matrices is computed:
```

:math:`\rho(t) = U(t)(\rho(0))`:

.. code-block::

    >>> rho_t = mesolve(H, fock_dm(2, 1), [0, 0.5, 1], c_ops=[sigmam()]).states
    >>> prop = propagator(H, [0, 0.5, 1], c_ops=[sigmam()])

    >>> print((rho_t[1] - prop[1](fock_dm(2, 1))).norm())
    7.23009476734681e-07

    >>> print((rho_t[2] - prop[2](fock_dm(2, 1))).norm())
    1.2666967766644768e-06

> The propagator is also available in class format:

Reviewer: ```suggestion
The propagator function is also available as a class:
```

.. code-block::

    >>> U = Propagator(H, c_ops=[sigmam()])

    >>> state_0_5 = U(0.5)(fock_dm(2, 1))
    >>> state_1 = U(1., t_start=0.5)(state_0_5)

    >>> print((rho_t[1] - state_0_5).norm())
    7.23009476734681e-07

    >>> print((rho_t[2] - state_1).norm())
    8.355518501351504e-07

The :obj:`.Propagator` can take ``options`` and ``args`` as a solver instance.

.. _propagator_solver:

> Using solver to compute propagator
> ==================================

Reviewer: ```suggestion
Using a solver to compute a propagator
=================================
```

> Many solver accept an operator as initial state input, when an identity matrix is
> passed, the propagator is computed, this can be used to compute propagator of
> Bloch-Redfield or Floquet equations:

Reviewer: ```suggestion
Many solvers accept an operator as the initial state. When an identity matrix is
passed as the initial state, the propagator is computed. This can be used to compute
a propagator for Bloch-Redfield or Floquet equations:
```
